File: frontend_verification/verify_app.py
import json
from playwright.sync_api import sync_playwright
import os

def run(playwright):
    browser = playwright.chromium.launch(headless=True)
    page = browser.new_page()

    # Get absolute path to index.html
    cwd = os.getcwd()
    file_path = f"file://{cwd}/frontend/index.html"

    print(f"Navigating to {file_path}")

    # Intercept API calls to mock backend
    def handle_route(route):
        request = route.request
        if "script.google.com" in request.url and request.method == "POST":
            # Check payload action
            try:
                post_data = json.loads(request.post_data)
                if post_data.get("action") == "getInitialData":
                    print("Mocking getInitialData")
                    route.fulfill(
                        status=200,
                        content_type="application/json",
                        body=json.dumps({
                            "rooms": ["Living Room", "Kitchen", "Bedroom"],
                            "items": [
                                { "ID": "1", "Room": "Living Room", "Type": "Main", "Name": "Sofa", "Price": 5000, "Dim_L": "200", "Dim_W": "90", "Dim_H": "80", "ImageID": "", "ProductURL": "" },
                                { "ID": "2", "Room": "Living Room", "Type": "Alternative", "ParentID": "1", "Name": "Sofa Option 2", "Price": 4500, "Dim_L": "190", "Dim_W": "85", "Dim_H": "80", "ImageID": "", "ProductURL": "" }
                            ]
                        })
                    )
                    return
            except Exception as e:
                print(f"Error parsing post data: {e}")

        # Continue other requests
        route.continue_()

    page.route("**/*", handle_route)

    page.goto(file_path)

    # 1. Verify Room List Rendered
    try:
        page.wait_for_selector(".room-card", timeout=5000)
        print("Room cards found.")
    except:
        print("Room cards NOT found. Saving error screenshot.")
        page.screenshot(path="frontend_verification/error_rooms.png")

    page.screenshot(path="frontend_verification/rooms_view.png")

    # 2. Click Room and Verify Detail
    page.click("text=Living Room")

    try:
        page.wait_for_selector(".item-card", timeout=2000)
        print("Item cards found.")
        # Expand Accordion
        page.click(".accordion-header")
        page.wait_for_timeout(500)
        page.screenshot(path="frontend_verification/room_detail_view.png")
    except:
         print("Item cards NOT found.")

    # 3. Verify Budget View
    page.click("[data-target='view-budget']")
    page.wait_for_timeout(500)
    page.screenshot(path="frontend_verification/budget_view.png")

    browser.close()

with sync_playwright() as playwright:
    run(playwright)

File: frontend_verification/verify_app.js
const { chromium } = require('playwright');
const path = require('path');

(async () => {
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  // Construct absolute file path to index.html
  const filePath = `file://${path.resolve('frontend/index.html')}`;
  console.log(`Navigating to: ${filePath}`);

  // Intercept the API call BEFORE navigation
  await page.route('**/*', async route => {
    const request = route.request();
    if (request.url().includes('script.google.com') && request.method() === 'POST') {
      const postData = JSON.parse(request.postData());

      if (postData.action === 'getInitialData') {
        console.log('Mocking getInitialData');
        await route.fulfill({
          contentType: 'application/json',
          body: JSON.stringify({
            rooms: ['Living Room', 'Kitchen', 'Bedroom'],
            items: [
              {
                ID: '1', Room: 'Living Room', Type: 'Main', Name: 'Sofa', Price: 5000,
                Dim_L: '200', Dim_W: '90', Dim_H: '80', ImageID: '', ProductURL: ''
              },
              {
                ID: '2', Room: 'Living Room', Type: 'Alternative', ParentID: '1', Name: 'Sofa Option 2', Price: 4500,
                Dim_L: '190', Dim_W: '85', Dim_H: '80', ImageID: '', ProductURL: ''
              }
            ]
          })
        });
        return;
      }
    }
    await route.continue();
  });

  await page.goto(filePath);

  // Wait for room list to appear (mock data loaded)
  try {
    await page.waitForSelector('.room-card', { timeout: 5000 });
    console.log('Room cards found');
  } catch (e) {
    console.log('Room cards not found within timeout');
    await page.screenshot({ path: 'frontend_verification/error_state.png' });
  }

  // Take screenshot of Rooms View
  await page.screenshot({ path: 'frontend_verification/rooms_view.png' });
  console.log('Screenshot saved: frontend_verification/rooms_view.png');

  // Click on "Living Room"
  const roomCard = await page.$('text=Living Room');
  if (roomCard) {
    await roomCard.click();
    console.log('Clicked Living Room');

    // Wait for details
    await page.waitForSelector('.item-card');

    // Open Accordion
    const accordionHeader = await page.$('.accordion-header');
    if (accordionHeader) {
      await accordionHeader.click();
      await page.waitForTimeout(500); // Animation
    }

    // Take screenshot of Room Detail View
    await page.screenshot({ path: 'frontend_verification/room_detail_view.png' });
    console.log('Screenshot saved: frontend_verification/room_detail_view.png');
  }

  // Switch to Budget View
  const budgetBtn = await page.$('[data-target="view-budget"]');
  if (budgetBtn) {
    await budgetBtn.click();
    await page.waitForTimeout(500);

    // Take screenshot of Budget View
    await page.screenshot({ path: 'frontend_verification/budget_view.png' });
    console.log('Screenshot saved: frontend_verification/budget_view.png');
  }

  await browser.close();
})();
